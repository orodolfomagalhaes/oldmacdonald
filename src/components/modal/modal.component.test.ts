import { expect, it, afterEach, beforeEach, describe, vi } from 'vitest';
import {ModalComponent} from "./modal.component.ts";

class MockModal {
  show = vi.fn();
  hide = vi.fn();
}

const mockBootstrap = {
  Modal: MockModal
};
const getCropTypesMock = [
  {
    "id": "ct001",
    "name": "Soja"
  },
  {
    "id": "ct002",
    "name": "Milho"
  },
  {
    "id": "ct003",
    "name": "Algodão"
  },
  {
    "id": "ct004",
    "name": "Feijão"
  },
  {
    "id": "ct005",
    "name": "Café"
  },
  {
    "id": "ct006",
    "name": "Arroz"
  },
  {
    "id": "ct007",
    "name": "Trigo"
  },
  {
    "id": "ct008",
    "name": "Cana-de-açúcar"
  },
  {
    "id": "ct009",
    "name": "Cevada"
  },
  {
    "id": "ct010",
    "name": "Aveia"
  }
];

const qs = (value: string) => document.querySelector(value);
const qsa = (value: string) => document.querySelectorAll(value);

// @ts-ignore
globalThis.bootstrap = mockBootstrap;

describe('ModalComponent', () => {
  let component: ModalComponent;

  const mockedFarmDataUpdate: any = {
    id: '789',
    farmName: 'Fazenda Esperança',
    landArea: 200,
    unitOfMeasure: 'hectare',
    address: 'Rua do Campo, 123, Interior - SP',
    cropProductions: [
      {
        id: 'cp001',
        cropTypeId: "ct001",
        isIrrigated: true,
        isInsured: false,
        farmId: '789'
      },
      {
        id: 'cp002',
        cropTypeId: "ct002",
        isIrrigated: true,
        isInsured: false,
        farmId: '789'
      }
    ]
  };

  globalThis.fetch = vi.fn();

  beforeEach(() => {
    component = new ModalComponent();
    document.body.appendChild(component);
  })

  afterEach(() => {
    document.body.removeChild(component);

    vi.clearAllMocks();
    vi.resetAllMocks();
  })

  it('should create component', () => {
    document.body.removeChild(component);
    const initFarmModalGeneralListenersSpy = vi.spyOn(component, 'initFarmModalGeneralListeners');
    document.body.appendChild(component);

    expect(component).toBeTruthy();
    expect(component).toMatchSnapshot();
    expect(initFarmModalGeneralListenersSpy).toHaveBeenCalledTimes(1)
  });

  it('should render modal with farm details when show-farm-detail event is called', () => {
    component.getCropTypes = vi.fn().mockReturnValue(getCropTypesMock);
    const spyUpdateTemplate = vi.spyOn(component, 'updateTemplate');
    const spyShowModal = vi.spyOn(component, 'showModal');
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);

    const farmName: any = qs('input[name="farmName"]');
    const landArea: any = qs('input[name="landArea"]');
    const unitOfMeasure: any = qs('select[name="unitOfMeasure"]');
    const address: any = qs('textarea[name="address"]');
    // @ts-ignore
    const formCropProductionOne = qsa('#production-list form')[0].elements;
    // @ts-ignore
    const formCropProductionTwo = qsa('#production-list form')[1].elements;

    // Farm details
    expect(farmName.value).toBe(mockedFarmDataUpdate.farmName);
    expect(Number(landArea.value)).toBe(mockedFarmDataUpdate.landArea);
    expect(unitOfMeasure.value).toBe(mockedFarmDataUpdate.unitOfMeasure);
    expect(address.value).toBe(mockedFarmDataUpdate.address);
    // Crop item one
    expect(formCropProductionOne.cropTypeId.value).toBe(mockedFarmDataUpdate.cropProductions[0].cropTypeId);
    expect(formCropProductionOne.isInsured.checked).toBe(mockedFarmDataUpdate.cropProductions[0].isInsured);
    expect(formCropProductionOne.isIrrigated.checked).toBe(mockedFarmDataUpdate.cropProductions[0].isIrrigated);
    // Crop item two
    expect(formCropProductionTwo.cropTypeId.value).toBe(mockedFarmDataUpdate.cropProductions[1].cropTypeId);
    expect(formCropProductionTwo.isInsured.checked).toBe(mockedFarmDataUpdate.cropProductions[1].isInsured);
    expect(formCropProductionTwo.isIrrigated.checked).toBe(mockedFarmDataUpdate.cropProductions[1].isIrrigated);

    expect(component.farm).toEqual(mockedFarmDataUpdate);
    expect(spyUpdateTemplate).toHaveBeenCalledWith(mockedFarmDataUpdate);
    expect(spyShowModal).toHaveBeenCalled();
    expect(component).toMatchSnapshot();
  });

  it('should open delete confirmation modal', () => {
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);

    const closeModalSpy = vi.spyOn(component, 'closeModal');
    const showModalSpy = vi.spyOn(component, 'showModal');
    const initDeleteConfirmationModalListenersSpy = vi.spyOn(component, 'initDeleteConfirmationModalListeners');
    const removeFarmModalListenersSpy = vi.spyOn(component, 'removeFarmModalListeners');

    const deleteButton: any = qs('button.delete');
    deleteButton?.click();

    expect(closeModalSpy).toHaveBeenCalledTimes(1);
    expect(closeModalSpy).toHaveBeenCalledWith('#farmModal');
    expect(showModalSpy).toHaveBeenCalledTimes(1);
    expect(showModalSpy).toHaveBeenCalledWith('#deleteConfirmationModal');
    expect(initDeleteConfirmationModalListenersSpy).toHaveBeenCalled();
    expect(removeFarmModalListenersSpy).toHaveBeenCalled();
  });

  it('should delete a farm when confirmed', async () => {
    // open modal
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);
    // open confirmation modal
    const deleteButton: any = qs('button.delete');
    deleteButton?.click();

    component.farmService.destroy = vi.fn().mockResolvedValue({deleted: true});

    const deleteSpy = vi.spyOn(component, 'delete');
    const handleResponseSpy = vi.spyOn(component, 'handleResponse');

    const confirmDeletionButton: any = qs('button.confirm');
    confirmDeletionButton?.click();

    expect(deleteSpy).toHaveBeenCalledTimes(1);
    expect(deleteSpy).toHaveBeenCalledWith(mockedFarmDataUpdate.id);

    await Promise.resolve()
    expect(handleResponseSpy).toHaveBeenCalled();

  });

  it('should go back to farm detail modal when back button is cliked', () => {
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);
    // open confirmation modal
    const deleteButton: any = qs('button.delete');
    deleteButton?.click();

    const closeModalSpy = vi.spyOn(component, 'closeModal');
    const showModalSpy = vi.spyOn(component, 'showModal');
    const removeDeleteConfirmationModalListenersSpy = vi.spyOn(component, 'removeDeleteConfirmationModalListeners');
    const initFarmModalListenersSpy = vi.spyOn(component, 'initFarmModalListeners');

    const back: any = qs('button.back');
    back?.click();

    expect(closeModalSpy).toHaveBeenCalledTimes(1);
    expect(closeModalSpy).toHaveBeenCalledWith('#deleteConfirmationModal');
    expect(showModalSpy).toHaveBeenCalledTimes(1);
    expect(showModalSpy).toHaveBeenCalledWith('#farmModal');
    expect(removeDeleteConfirmationModalListenersSpy).toHaveBeenCalled();
    expect(initFarmModalListenersSpy).toHaveBeenCalled();
  });

  it('should reset farm object', () => {
    const resetedFarm = {id: '',farmName: '',landArea: 0,unitOfMeasure: 'hectare',address: '',cropProductions: []}
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);
    expect(component.farm).toEqual(mockedFarmDataUpdate);

    component.farm = component.resetFarm();
    expect(component.farm).toEqual(resetedFarm);
  });

  it('should add new crop item to list', () => {
    const spyInitFarmModalListeners = vi.spyOn(component, 'initFarmModalListeners');
    const spyAddCropProductionCard = vi.spyOn(component, 'addCropProductionCard');

    component.showModal('#farmModal');
    const button: any = qs('.add-production');
    button?.click();
    let productionList: any = qsa('#production-list .card');

    expect(spyInitFarmModalListeners).toHaveBeenCalled();
    expect(spyAddCropProductionCard).toHaveBeenCalled();
    expect(productionList.length).toBe(1);

    button?.click();
    productionList = qsa('#production-list .card');
    expect(productionList.length).toBe(2);
  });

  it('should remove crop item from list', () => {
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);

    let productionList: any = qsa('#production-list .card');
    expect(productionList.length).toBe(2);

    const removeButton: any = qs('#production-list .card .remove');
    removeButton?.click();
    productionList = qsa('#production-list .card');
    expect(productionList.length).toBe(1);

    // @ts-ignore
    expect(component.farm.cropProductions[0]).toHaveProperty('deletedAt');
  });

  it('should save an updated farm', async () => {
    const event = new CustomEvent('show-farm-detail', {detail: mockedFarmDataUpdate, bubbles: true});
    document.dispatchEvent(event);
    const farmName: any = qs('input[name="farmName"]');
    farmName.value = 'Fazenda Bom Sucesso';
    const farmNameEvent = new Event('input', { bubbles: true });
    farmName.dispatchEvent(farmNameEvent);
    const submitButton: any = qs('button.submit');
    submitButton?.click();
    component.farmService.save = vi.fn().mockResolvedValue({success: true});

    const validateFormSpy = vi.spyOn(component, 'validateForm');
    component.save();

    const farmForm: any = qs('.modal-body form');
    expect(validateFormSpy).toHaveBeenCalledTimes(1);
    expect(validateFormSpy).toHaveBeenNthCalledWith(1, farmForm);

    const handleResponseSpy = vi.spyOn(component, 'handleResponse');
    await Promise.resolve();
    expect(handleResponseSpy).toHaveBeenCalled();
  });

  it('should validate an invalid field and mark it with is-invalid class', () => {
    const event = new CustomEvent('add-new-farm', {bubbles: true});
    document.dispatchEvent(event);

    const farmName: any = qs('input[name="farmName"]');
    const landArea: any = qs('input[name="landArea"]');
    const productionsElement: any = qs('#productions');

    const validateFormSpy = vi.spyOn(component, 'validateForm');
    const farmForm: any = qs('.modal-body form');
    component.save();

    expect(validateFormSpy).toHaveBeenCalledTimes(1);
    expect(validateFormSpy).toHaveBeenNthCalledWith(1, farmForm);
    expect(validateFormSpy).toReturnWith(false);
    expect(farmName.classList.contains('is-invalid')).toBeTruthy();
    expect(landArea.classList.contains('is-invalid')).toBeTruthy();
    expect(productionsElement.classList.contains('is-invalid')).toBeTruthy();
  });

  it('should fill the invalid fields and remove is-invalid class', async () => {
    const event = new CustomEvent('add-new-farm', {bubbles: true});
    component.farmService.create = vi.fn().mockResolvedValue({success: true});
    document.dispatchEvent(event);
    component.showModal('#farmModal');
    const form: any = qs('.modal-body form');
    form.elements.landArea.value = '';
    component.save(); // to make it invalid
    const productionsElement: any = qs('#productions');
    // update values
    form.elements.farmName.value = 'Fazenda Bom Sucesso';
    form.elements.farmName.dispatchEvent(new Event('input', { bubbles: true }));
    form.elements.landArea.value = '1200';
    form.elements.landArea.dispatchEvent(new Event('input', { bubbles: true }));
    const addProductionButton: any = qs('.add-production');
    addProductionButton?.click();

    const validateFormSpy = vi.spyOn(component, 'validateForm');
    component.save();

    expect(validateFormSpy).toHaveBeenCalledTimes(1);
    expect(validateFormSpy).toHaveBeenNthCalledWith(1, form);
    expect(validateFormSpy).toReturnWith(true);
    expect(form.elements.farmName.classList.contains('is-invalid')).toBeFalsy();
    expect(form.elements.landArea.classList.contains('is-invalid')).toBeFalsy();
    expect(productionsElement.classList.contains('is-invalid')).toBeFalsy();

    const handleResponseSpy = vi.spyOn(component, 'handleResponse');
    await Promise.resolve();
    expect(handleResponseSpy).toHaveBeenCalled();
  });

  it('should get CropTypes list', () => {
    const getCropTypes = vi.spyOn(component, 'getCropTypes');
    component.getCropTypes();
    expect(getCropTypes).toHaveBeenCalled();
    expect(getCropTypes).toReturnWith([]);
  });

  it('should close modal when event is fired', () => {
    const closeModalEvent = new Event('hidden.bs.modal', { bubbles: true });
    const closeButton: any = qs('.btn-close')
    closeButton.dispatchEvent(closeModalEvent);
  });

  it('should handle the response of farmService requests', () => {
    const successResponse = { success: true, message: 'Success!' };
    const errorResponse = { success: false, message: 'Error!' };
    const handleResponseSpy = vi.spyOn(component, 'handleResponse');
    const showToastEventSpy = vi.spyOn(component, 'showToastEvent');

    const showToastEvent = vi.fn();
    document.addEventListener('show-toast', showToastEvent);

    component.handleResponse(successResponse);
    expect(handleResponseSpy).toHaveBeenCalledWith(successResponse);
    component.handleResponse(errorResponse);
    expect(handleResponseSpy).toHaveBeenCalledWith(errorResponse);
    expect(showToastEventSpy).toHaveBeenCalledTimes(2);
    expect(showToastEvent).toHaveBeenCalledTimes(2);
  });

});