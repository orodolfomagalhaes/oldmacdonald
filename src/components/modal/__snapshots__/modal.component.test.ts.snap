// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalComponent > should create component 1`] = `
<modal-component>
  
      
  <div
    aria-hidden="true"
    aria-labelledby="farmModalLabel"
    class="modal"
    id="farmModal"
    tabindex="-1"
  >
    
        
    <div
      class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
    >
      
          
      <div
        class="modal-content"
      >
        
          
            
        <div
          class="modal-header"
        >
          
              
          <h5
            class="modal-title"
          >
            Farm Registration
          </h5>
          
              
          <button
            aria-label="Close modal"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
          
            
        </div>
        
            
            
        <div
          class="modal-body"
        >
          
              
          <form>
            
                
            <div
              class="mb-3"
            >
              
                  
              <label
                class="form-label"
              >
                Farm Name *
              </label>
              
                  
              <input
                aria-required="true"
                class="form-control"
                name="farmName"
                required=""
                type="text"
                value=""
              />
              
                
            </div>
            
                
            <div
              class="row mb-3"
            >
              
                  
              <div
                class="col-md-6"
              >
                
                    
                <label
                  class="form-label"
                >
                  Land Area *
                </label>
                
                    
                <input
                  aria-required="true"
                  class="form-control"
                  name="landArea"
                  required=""
                  type="number"
                  value="0"
                />
                
                  
              </div>
              
                  
              <div
                class="col-md-6"
              >
                
                    
                <label
                  class="form-label"
                >
                  Unit of Measure *
                </label>
                
                    
                <select
                  class="form-select"
                  name="unitOfMeasure"
                >
                  
                      
                  <option
                    selected=""
                    value="hectare"
                  >
                    Hectare
                  </option>
                  
                      
                  <option
                    value="acre"
                  >
                    Acre
                  </option>
                  
                    
                </select>
                
                  
              </div>
              
                
            </div>
            
                
            <div
              class="mb-5"
            >
              
                  
              <label
                class="form-label"
              >
                Address
              </label>
              
                  
              <textarea
                class="form-control"
                name="address"
              />
              
                
            </div>
                        
              
          </form>
          

              
          <div
            class="mb-3"
          >
            
                  
            <div
              class="d-flex justify-content-between align-items-center mb-2"
              id="productions"
            >
              
                    
              <h6>
                Crop Productions *
              </h6>
              
                    
              <button
                class="btn btn-sm btn-primary add-production"
                type="button"
              >
                
                      Add Production
                    
              </button>
              
                  
            </div>
            

                  
            <div
              id="production-list"
            />
            
                
          </div>
                          
            
        </div>
        
            
            
        <div
          class="modal-footer "
        >
          
                          
              
          <div
            class="main-buttons"
          >
            
                
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
              
                  Cancel
                
            </button>
            
                
            <button
              class="btn btn-primary submit"
              type="submit"
            >
              
                  Save
                
            </button>
            
              
          </div>
          
              
            
        </div>
        
          
      </div>
      
        
    </div>
    
      
  </div>
  
      
  <div
    aria-hidden="true"
    aria-labelledby="deleteConfirmationModalLabel"
    class="modal"
    id="deleteConfirmationModal"
    tabindex="-1"
  >
    
        
    <div
      class="modal-dialog modal-dialog-centered"
    >
      
          
      <div
        class="modal-content"
      >
        
            
        <div
          class="modal-header"
        >
          
              
          <h1
            class="modal-title fs-5"
            id="deleteConfirmationModalLabel"
          >
            Confirm deletion
          </h1>
          
              
          <button
            aria-label="Close modal"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
          
            
        </div>
        
            
        <div
          class="modal-body"
        >
          
              Are you sure you want to delete this farm?
            
        </div>
        
            
        <div
          class="modal-footer"
        >
          
              
          <button
            aria-label="Cancel deletion and return to the farm details modal"
            class="btn btn-secondary back"
          >
            Back
          </button>
          
              
          <button
            aria-label="Confirm deletion of the farm"
            class="btn btn-danger confirm"
          >
            Do it!
          </button>
          
            
        </div>
        
          
      </div>
      
        
    </div>
    
      
  </div>
</modal-component>
`;

exports[`ModalComponent > should render modal with farm details when show-farm-detail event is called 1`] = `
<modal-component>
  
      
  <div
    aria-hidden="true"
    aria-labelledby="farmModalLabel"
    class="modal"
    id="farmModal"
    tabindex="-1"
  >
    
        
    <div
      class="modal-dialog modal-lg modal-dialog-centered modal-dialog-scrollable"
    >
      
          
      <div
        class="modal-content"
      >
        
          
            
        <div
          class="modal-header"
        >
          
              
          <h5
            class="modal-title"
          >
            Farm Registration
          </h5>
          
              
          <button
            aria-label="Close modal"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
          
            
        </div>
        
            
            
        <div
          class="modal-body"
        >
          
              
          <form>
            
                
            <div
              class="mb-3"
            >
              
                  
              <label
                class="form-label"
              >
                Farm Name *
              </label>
              
                  
              <input
                aria-required="true"
                class="form-control"
                name="farmName"
                required=""
                type="text"
                value="Fazenda EsperanÃ§a"
              />
              
                
            </div>
            
                
            <div
              class="row mb-3"
            >
              
                  
              <div
                class="col-md-6"
              >
                
                    
                <label
                  class="form-label"
                >
                  Land Area *
                </label>
                
                    
                <input
                  aria-required="true"
                  class="form-control"
                  name="landArea"
                  required=""
                  type="number"
                  value="200"
                />
                
                  
              </div>
              
                  
              <div
                class="col-md-6"
              >
                
                    
                <label
                  class="form-label"
                >
                  Unit of Measure *
                </label>
                
                    
                <select
                  class="form-select"
                  name="unitOfMeasure"
                >
                  
                      
                  <option
                    selected=""
                    value="hectare"
                  >
                    Hectare
                  </option>
                  
                      
                  <option
                    value="acre"
                  >
                    Acre
                  </option>
                  
                    
                </select>
                
                  
              </div>
              
                
            </div>
            
                
            <div
              class="mb-5"
            >
              
                  
              <label
                class="form-label"
              >
                Address
              </label>
              
                  
              <textarea
                class="form-control"
                name="address"
              >
                Rua do Campo, 123, Interior - SP
              </textarea>
              
                
            </div>
                        
              
          </form>
          

              
          <div
            class="mb-3"
          >
            
                  
            <div
              class="d-flex justify-content-between align-items-center mb-2"
              id="productions"
            >
              
                    
              <h6>
                Crop Productions *
              </h6>
              
                    
              <button
                class="btn btn-sm btn-primary add-production"
                type="button"
              >
                
                      Add Production
                    
              </button>
              
                  
            </div>
            

                  
            <div
              id="production-list"
            >
              <div
                class="card mb-2"
              >
                <div
                  class="card-body"
                >
                  
        
                  <form
                    id="cp001"
                  >
                    
          
                    <div
                      class="row"
                    >
                      
            
                      <div
                        class="col-md-6"
                      >
                        
              
                        <label
                          class="form-label"
                        >
                          Crop Type
                        </label>
                        
              
                        <select
                          class="form-select"
                          name="cropTypeId"
                        >
                          
                
                  
                          <option
                            selected=""
                            value="ct001"
                          >
                            
                    Soja
                  
                          </option>
                          
                
                  
                          <option
                            value="ct002"
                          >
                            
                    Milho
                  
                          </option>
                          
                
                  
                          <option
                            value="ct003"
                          >
                            
                    AlgodÃ£o
                  
                          </option>
                          
                
                  
                          <option
                            value="ct004"
                          >
                            
                    FeijÃ£o
                  
                          </option>
                          
                
                  
                          <option
                            value="ct005"
                          >
                            
                    CafÃ©
                  
                          </option>
                          
                
                  
                          <option
                            value="ct006"
                          >
                            
                    Arroz
                  
                          </option>
                          
                
                  
                          <option
                            value="ct007"
                          >
                            
                    Trigo
                  
                          </option>
                          
                
                  
                          <option
                            value="ct008"
                          >
                            
                    Cana-de-aÃ§Ãºcar
                  
                          </option>
                          
                
                  
                          <option
                            value="ct009"
                          >
                            
                    Cevada
                  
                          </option>
                          
                
                  
                          <option
                            value="ct010"
                          >
                            
                    Aveia
                  
                          </option>
                          
                
              
                        </select>
                        
            
                      </div>
                      
            
                      <div
                        class="col-md-6"
                      >
                        
              
                        <div
                          class="form-check mt-4"
                        >
                          
                
                          <input
                            checked=""
                            class="form-check-input"
                            name="isIrrigated"
                            type="checkbox"
                          />
                          
                
                          <label
                            class="form-check-label"
                          >
                            Irrigated
                          </label>
                          
              
                        </div>
                        
              
                        <div
                          class="form-check"
                        >
                          
                
                          <input
                            class="form-check-input"
                            name="isInsured"
                            type="checkbox"
                          />
                          
                
                          <label
                            class="form-check-label"
                          >
                            Insured
                          </label>
                          
              
                        </div>
                        
            
                      </div>
                      
        
                    </div>
                    
        
                  </form>
                  
        
                  <button
                    class="btn btn-sm btn-danger mt-2 remove"
                    type="button"
                  >
                    
          Remove
        
                  </button>
                  
      
                </div>
              </div>
              <div
                class="card mb-2"
              >
                <div
                  class="card-body"
                >
                  
        
                  <form
                    id="cp002"
                  >
                    
          
                    <div
                      class="row"
                    >
                      
            
                      <div
                        class="col-md-6"
                      >
                        
              
                        <label
                          class="form-label"
                        >
                          Crop Type
                        </label>
                        
              
                        <select
                          class="form-select"
                          name="cropTypeId"
                        >
                          
                
                  
                          <option
                            value="ct001"
                          >
                            
                    Soja
                  
                          </option>
                          
                
                  
                          <option
                            selected=""
                            value="ct002"
                          >
                            
                    Milho
                  
                          </option>
                          
                
                  
                          <option
                            value="ct003"
                          >
                            
                    AlgodÃ£o
                  
                          </option>
                          
                
                  
                          <option
                            value="ct004"
                          >
                            
                    FeijÃ£o
                  
                          </option>
                          
                
                  
                          <option
                            value="ct005"
                          >
                            
                    CafÃ©
                  
                          </option>
                          
                
                  
                          <option
                            value="ct006"
                          >
                            
                    Arroz
                  
                          </option>
                          
                
                  
                          <option
                            value="ct007"
                          >
                            
                    Trigo
                  
                          </option>
                          
                
                  
                          <option
                            value="ct008"
                          >
                            
                    Cana-de-aÃ§Ãºcar
                  
                          </option>
                          
                
                  
                          <option
                            value="ct009"
                          >
                            
                    Cevada
                  
                          </option>
                          
                
                  
                          <option
                            value="ct010"
                          >
                            
                    Aveia
                  
                          </option>
                          
                
              
                        </select>
                        
            
                      </div>
                      
            
                      <div
                        class="col-md-6"
                      >
                        
              
                        <div
                          class="form-check mt-4"
                        >
                          
                
                          <input
                            checked=""
                            class="form-check-input"
                            name="isIrrigated"
                            type="checkbox"
                          />
                          
                
                          <label
                            class="form-check-label"
                          >
                            Irrigated
                          </label>
                          
              
                        </div>
                        
              
                        <div
                          class="form-check"
                        >
                          
                
                          <input
                            class="form-check-input"
                            name="isInsured"
                            type="checkbox"
                          />
                          
                
                          <label
                            class="form-check-label"
                          >
                            Insured
                          </label>
                          
              
                        </div>
                        
            
                      </div>
                      
        
                    </div>
                    
        
                  </form>
                  
        
                  <button
                    class="btn btn-sm btn-danger mt-2 remove"
                    type="button"
                  >
                    
          Remove
        
                  </button>
                  
      
                </div>
              </div>
            </div>
            
                
          </div>
                          
            
        </div>
        
            
            
        <div
          class="modal-footer justify-content-between"
        >
          
              
          <button
            class="btn btn-danger delete"
            type="button"
          >
            
                        Delete Farm
                      
          </button>
                      
              
          <div
            class="main-buttons"
          >
            
                
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
              
                  Cancel
                
            </button>
            
                
            <button
              class="btn btn-primary submit"
              type="submit"
            >
              
                  Save
                
            </button>
            
              
          </div>
          
              
            
        </div>
        
          
      </div>
      
        
    </div>
    
      
  </div>
  
      
  <div
    aria-hidden="true"
    aria-labelledby="deleteConfirmationModalLabel"
    class="modal"
    id="deleteConfirmationModal"
    tabindex="-1"
  >
    
        
    <div
      class="modal-dialog modal-dialog-centered"
    >
      
          
      <div
        class="modal-content"
      >
        
            
        <div
          class="modal-header"
        >
          
              
          <h1
            class="modal-title fs-5"
            id="deleteConfirmationModalLabel"
          >
            Confirm deletion
          </h1>
          
              
          <button
            aria-label="Close modal"
            class="btn-close"
            data-bs-dismiss="modal"
            type="button"
          />
          
            
        </div>
        
            
        <div
          class="modal-body"
        >
          
              Are you sure you want to delete this farm?
            
        </div>
        
            
        <div
          class="modal-footer"
        >
          
              
          <button
            aria-label="Cancel deletion and return to the farm details modal"
            class="btn btn-secondary back"
          >
            Back
          </button>
          
              
          <button
            aria-label="Confirm deletion of the farm"
            class="btn btn-danger confirm"
          >
            Do it!
          </button>
          
            
        </div>
        
          
      </div>
      
        
    </div>
    
      
  </div>
</modal-component>
`;
